### ISSUE TITLE
Require input bed file to have locus ids as well?

### ISSUE THREAD

**Issue by gwct** [harvardinformatics/PhyloAcc-interface#1]
Currently PhyloAcc requires both a bed file with locus coordinates and an ID file with locus names. [bed format standards](https://genome-asia.ucsc.edu/FAQ/FAQformat.html#format1) indicate the fourth column is reserved for a name for the region specified on that line, however this field is optional. Should we require input bed files to have this column to obviate the need for an input ID file as well?

If so, we could make this change only on the front-end, which would then extract the IDs and write a file, or we could also change this requirement for the main PhyloAcc codebase and get rid of the ID file completely.

**Comment by HanY-H** [comment]
I am not sure if I understand your question completely, but I feel the ID file will be needed. the BED file tells the program where to locate the sequence of each element in FASTA file, and also gives an ID to each element. the FASTA file may contain the sequence of one whole chromosome. But an user may only want to run certain elements on the chromosome. They specify these element IDs in the ID file.

**Comment by gwct** [comment]
I guess I was thinking that the user would only include the elements they wanted to run in the bed file, in which case I don't think there's a need for the ID file. But are there use cases I'm not thinking of where they wouldn't want to remove unwanted elements in the bed file for a given run?

**Comment by tsackton** [comment]
I think the use case here would be if, for example, someone downloaded a bed file of conserved elements but only wanted to analyze the ones that overlapped ATAC-seq peaks or something like that. Of course the user could just filter the bed file to produce a file for analysis but I think we should be flexible with the program inputs to allow analysis of a subset of the bed file.

However, I think it might make sense to handle this in the following way. Have the Python front end require a bed file with names, and, optionally, a file of IDs to analyze if the user does not want to analyze every element in the bed file. If the ID file is not supplied we assume that all elements in the BED file will be analyzed. 

Does this seem reasonable to everyone?

**Comment by xyz111131** [comment]
yes the ID file is optional for the C++ code, in case the user only wants to run a subset of the elements. I think Phast (and other programs) also has options to filter for some elements to run.

**Comment by gwct** [comment]
Ok, that's great, I didn't realize it was optional already. Tim's suggestion that the default behavior would be to run all regions in the bed file, but if an ID file was provided we will limit it to those regions sounded reasonable to me as well, so I guess we'll keep it like that in the python code as well.

**Comment by gwct** [comment]
There are still some issues about the ID file, at least for the gene tree version. 

1. An ID file is still necessary -- the program crashes if one is not provided. Since batching with the interface implies that all elements in a batch will be run, we should be able to submit the batch without an ID file.
2. IDs must be integers. This prevents descriptive IDs such as "chr1-element19" or something.
3. PhyloAcc generates its own internal integer ID rather than using the ID provided in the 4th column of the bed file.

We can technically work around all of these issues with the interface. We can provide an ID file with all elements (1) and we can keep track of how IDs are linked (2 and 3). But resolving these would make things a bit more streamlined.

**Comment by xyz111131** [comment]
1. It's weird. Could you point to the main.cpp in PhyloAcc-GT you are using?
2. Yes, currently PhyloAcc can only use the order but not element ID. 
3. PhyloAcc takes the order of each element in the input bed as the ID.

**Comment by gwct** [comment]
For the crashing, the folder with the binary and source I'm using is here:

`/n/home07/gthomas/projects/PhyloAcc-interface/testSRC_debug4_2tree`

The behavior is inconsistent. Sometimes it crashes right away with a Segmentation fault, sometimes it runs for a second before an error (`*** Error in '/n/home07/gthomas/projects/PhyloAcc-interface/testSRC_debug4_2tree/PhyloAcc-GT_piQ': double free or corruption (out): 0x00002b9fc41bfc30 ***`) and a long backtrace and memory map, and sometimes it runs but is clearly having problems in the output (see logfile: `/n/home07/gthomas/projects/PhyloAcc/src/PhyloAcc-interface/id-test/without-id-output/without-id.log`)

Happy to help debug if needed.

**Comment by HanY-H** [comment]
Hi Gregg,

Can you send me the err log file you mentioned below? I don’t have access to your directory. 

In main.cpp, if ID file is not supplied, and batch = -1 (out default value), it will generate sequence 0 - 499 as IDs, and compute the first 500 elements in bed file. 
Do you have a parameter called "batch" in your interface, that may intervene here?
How many elements do you have in the test bed file? Is it possible the job terminated/corrupted when it tries to run an element with ID >number of elements/lines in bed file?

Best,
Han

**Comment by xyz111131** [comment]
If that is the case, then the main.cpp in phyloacc-GT, should be changed
to:
https://github.com/phyloacc/PhyloAcc/blob/ac4e66cd3cbaf388f724dfc1c796c28498173e42/src/PhyloAcc-ST/main.cpp#L352


which will run all the element in the bed.

HanY-H ***@***.***> 于2022年2月17日周四 08:26写道：

> Hi Gregg,
>
> Can you send me the err log file you mentioned below? I don’t have access
> to your directory.
>
> In main.cpp, if ID file is not supplied, and batch = -1 (out default
> value), it will generate sequence 0 - 499 as IDs, and compute the first 500
> elements in bed file.
> Do you have a parameter called "batch" in your interface, that may
> intervene here?
> How many elements do you have in the test bed file? Is it possible the job
> terminated/corrupted when it tries to run an element with ID >number of
> elements/lines in bed file?
>
> Best,
> Han
>
>
